### ISSUE TITLE
Missing documentation for internal mesh data structure

### ISSUE THREAD

**Issue by burnpanck** [sfepy/sfepy#746]
As an extension to #731, I'm now struggling to create surface region programmatically, based off a set of faces, as specified by their connectivity to the vertices.

More generally, there is no clear documentation on how the various indices used in the mesh and region representation relate to each other. For example, the `Region.from_facets` specified in it's docstring the following:
```
facets : array
    The array with indices to unique facets.
```
However, there is absolutely no indication what the meaning of those indices is.

I have attempted to follow the code all the way through `cmesh.pyx` and `mesh.h/.c`, and understand now that the cmesh internally stores direct mappings from indices of entities of all dimension to indices of all adjacent entities of all dimensions (the "connectivity"). However, the fact that there is a method `CMesh.get_local_ids` indicates that there are "local" and "global" ids, and I haven't figured out yet what these refer to.

To summarise, I believe it would be helpful to have a clear documentation of internal structure and API guarantees of Meshes and Regions - not just for creating meshes, but also to give context to the documentation e.g. of `Functions` which are passed coordinates of quadrature points - again currently lacking any documentation about their indexing.

If someone can help me out with a few pointers especially around the "virtual" entities in a mesh (that is neither cells nor vertices), e.g. how to determine the vertices corresponding to the facet ids of a region and vice versa, I might be able to add a bit of detail to the docstrings to help further explorers of the code.

**Comment by rc** [comment]
Essentially, the code in `mesh.c` and subsequently `cmesh.pyx` implements/adapts the algorithms published in https://arxiv.org/abs/1205.3081.

Basically, a 3D volumetric mesh is constructed using vertices (entities of topological dimension 0) with given coordinates and cells (topological dimension 3), defined using  a connectivity of the vertices (= the classical connectivity, returned by `CMesh.get_conn(3, 0)`).  The algorithms allow computing any other connectivity, e.g. cells connected to vertices (`CMesh.get_conn(0, 3)`), edges of faces, faces of cells, ... Those can then be used in regions to compute the region entities (cells, facets (faces), edges, vertices, depending on the region kind - facet region has no cells etc.). So a region contains those global indices of the mesh entities, as defined in the CMesh. For details see the referenced article.

The `CMesh.get_local_ids()` is then used in connection with `CMesh.get_incident()`. For example, let's have a facet region. For certain calculations, we need to get the cells corresponding to the region facets (<- `CMesh.get_incident()`) but also the local indices of the facets within those cells `<- CMesh.get_local_ids()`, see `Region.get_facet_indices()`. So if a cell X has, in its facets connectivity, facets [A, B, C, D], where X, A, B, C, D are global ids from the CMesh, the local indices of A, B, C, D in X are 0, 1, 2, 3. In a cell Y with facets [E, F, A, G] the facet A, common with the cell X, has now the local index 2. The ordering of lower dimensional entities in a cell is given by `geometry_data` in `sfepy/discrete/fem/geometry_element.py`.

Those are rather low level layers of the code, you might be the first to dig so deep :) In any case, it would be great to have it documented. I will probably add the above explanation somewhere into the docs (developer's guide), as having a comprehensive/broader context description would complement the (to be improved) docstrings.

Now to answer your question about using `Region.from_facets()`: the `facets` are the global indices as given e.g. by `CMesh.get_conn(3, 2)`, that is, given by the `CMesh` connectivities. Note that `CMesh.setup_connectivity()` has to be called to have a particular connectivity computed/accessible (shameless segfault in `CMesh.get_conn()` if it does not exist).

Does it help? Again, feel free to post a self-contained example so that I could run it.

**Comment by rc** [comment]
Actually, you can also visualize the entities and their numberings using:
```
python3 script/plot_mesh.py meshes/elements/2_4_2.mesh 
```
I have just found that 3D plots are broken now, and it is usable only for very small meshes, but it is enough to get the idea - Look at the image and also at the mesh connectivity in the file. The global numbers are on the entities, the cell-local ones are inside the cells next to each entity.
